add kep-115 Subgroup support (#127)
kubernetes-sigs/lws: kep/115-Subgroup-support/README.md

LeaderWorkerSet does not currently support having a group for each phase.

## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation. What is the desired outcome and how do we measure success?.
The "Design Details" section below is for the real
nitty-gritty.
-->

### User Stories (Optional)

<!--
Detail the things that people will be able to do if this KEP is implemented.
Include as much detail as possible so that people can understand the "how" of
the system. The goal here is to make this feel real for users without getting
bogged down.
-->

#### Story 1

> As a user, I should be able to have smaller groups within my LeaderWorkerSet deployment and be able to deploy each subgroup on a different topology. 

Author: fixed

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->
To avoid the changes from affecting initial functionality and to not set 
unnecessary labels, there will be no default behavior if `SubGroupSize` is not 
set

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->
### LeaderWorkerSet API
We extend the LeaderWorkerSet API to introduce a new field: subGroupSize to opt in and set the number of pods that each subgroup will contain. Current behavior is kept if not set. 

```
type LeaderWorkerSetSpec struct {

> 	 // Number of pods per subgroup

Reviewer: pls document the the constraints on this, 1) it must not be greater than size, 2) size must be divisible by this value 3) it is immutable

Author: Added

### Implementation
- Two new annotations will be added
  - `leaderworkerset.gke.io/exclusive-topology-subgroup `
  - `leaderworkerset.gke.io/subgroup-size` 
- Three new labels will be added,
  - `leaderworkerset.gke.io/subgroup-index = worker-index/subGroupSize`
    - Tracks which subgroup the pod is part of 
  - `leaderworkerset.gke.io/subgroup-worker-index = worker-index%subGroupSize`
    - index/identity of a pod inside the pod's subgroup
  - `leaderworkerset.gke.io/subgroup-key` 
    - Pods that are part of the same subgroup will have an annotation that is a unique hash value will be generated from the name of the leader, and the subgroup-index

To support exclusive placement at the subgroup level, the pod webhook will inject the new labels, and set the pod affinity/anti-affinity on all the pods. If both levels of pod affinity are set, then the leader pod will contain two pod affinities, while the workers will have a node selector, and a single pod affinity set. 

### TPU Environment Variable Injection

> Because the value of the TPU environmental variables will vary between subgroups, the way they are injected will be extended. `TPU_WOKRER_ID` will be the value of `subgroup-worker-index`, while the value of `TPU_HOSTNAMES` will only be a list of the pods in the same subgroup.

Reviewer: ```suggestion
Because the value of the TPU environmental variables will vary between subgroups, the way they are injected will be extended. `TPU_WORKER_ID` will be the value of `subgroup-worker-index`, while the value of `TPU_HOSTNAMES` will only be a list of the pods in the same subgroup.
```

Author: fixed

> LeaderWorkerSet supports the leader not requesting TPU resources. This raises a problem when determining the values of `subgroup-index` and s`ubgroup-worker-index`. Suppose we have two TPU slices with two hosts each. Since the leader doesn’t request TPU resources, there will be four workers + the leader, meaning that one of the workers will have a worker index of four. Because of the way the new labels are calculated, this worker will have a subgroup-index of two, creating three subgroup indices (0,1,2) even though there are only two TPU slices.

Reviewer: ```suggestion
LeaderWorkerSet supports the leader not requesting TPU resources. This raises a problem when determining the values of `subgroup-index` and `subgroup-worker-index`. Suppose we have two TPU slices with two hosts each. Since the leader doesn’t request TPU resources, there will be four workers + the leader, meaning that one of the workers will have a worker index of four. Because of the way the new labels are calculated, this worker will have a subgroup-index of two, creating three subgroup indices (0,1,2) even though there are only two TPU slices.
```

Author: fixed

      - [Integration tests](#integration-tests)
  - [Alternatives](#alternatives)
    - [Only Set Pod-Affinity on workers with `subgroup-worker-index=0`](#only-set-pod-affinity-on-workers-with-subgroup-worker-index0)
<!-- /toc -->

## Motivation

<!--
This section is for explicitly listing the motivation, goals, and non-goals of
this KEP.  Describe why the change is important and the benefits to users. The
motivation section can optionally provide links to [experience reports] to
demonstrate the interest in a KEP within the wider Kubernetes community.

[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
-->

Disaggregated serving is an optimization made for LLM inference workloads. It takes advantage of the fact that the two phases of inference have different characteristics and thus it can be beneficial to run them on different machines. State of the art LLM serving frameworks such as [vLLM](https://github.com/vllm-project/vllm/issues/2472) are already adding support for this optimization based on the paper released by [Microsoft](https://www.microsoft.com/en-us/research/publication/splitwise-efficient-generative-llm-inference-using-phase-splitting/). 

> LeaderWorkerSet does not currently support having a group for each phase.

Reviewer 2: ```suggestion
LeaderWorkerSet does not currently support having a group for each phase.
```
This KEP is to have LeaderWorkerSet to support running both prefill/decode servers in one pod group (aka one replica)

Author: Fixed

set

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->
### LeaderWorkerSet API
We extend the LeaderWorkerSet API to introduce a new field: subGroupSize to opt in and set the number of pods that each subgroup will contain. Current behavior is kept if not set. 

```
type LeaderWorkerSetSpec struct {
	 // Number of pods per subgroup. This value is immutable,
   // and must not be greater than LeaderWorkerSet.Spec.Size.
	SubGroupSize *int32 'json:"subGroupSize,omitempty'
} 
```

### Exclusive Topology Support
LeaderWorkerSet can guarantee that the leader and the workers are placed in the same topology if the `leaderworkerset.gke.io/exclusive-topology` annotation is set. Similarly, we will support that the pods within the same subgroup will be placed in the same topology with a new annotation `leaderworkerset.gke.io/exclusive-topology-subgroup`, so that the new changes can support up to two levels of pod affinity. 

> 

Reviewer 2: We could give an example for two levels of exclusive placement and describe how pods will be scheduled at last.

Reviewer 3: +1 to have an example.

Author: added example

If the leader does not request TPU resources, then the labels will have the following values
`leaderworkerset.gke.io/subgroup-index = (workerIndex - 1) / subGroupSize`
`leaderworkerset.gke.io/subgroup-worker-index = (workerIndex - 1) % subGroupSize`

### Test Plan

<!--
**Note:** *Not required until targeted at a release.*
The goal is to ensure that we don't accept enhancements with inadequate testing.

All code is expected to have adequate tests (eventually with coverage
expectations). Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
when drafting this test plan.

[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
-->

> [ ] I/we understand the owners of the involved components may require updates to

Reviewer 2: qq, what is the "[ ]" at the start of the line?

Author: A checkbox which is part of the template

However, if complete unit test coverage is not possible, explain the reason of it
together with explanation why this is acceptable.
-->

<!--
Additionally, try to enumerate the core package you will be touching
to implement this enhancement and provide the current unit coverage for those
in the form of:
- <package>: <date> - <current test coverage>

This can inform certain test coverage improvements that we want to do before
extending the production code to implement this enhancement.
-->

Will be added to ensure that the annotations are injected into the leader and worker pods, and to verify that the new values of TPU hostnames and TPU worker ID are injected. 

#### Integration tests

<!--
Describe what tests will be added to ensure proper quality of the enhancement.

After the implementation PR is merged, add the names of the tests here.
-->

- The pod-webhook integration tests should test
  - Pod Affinity/Anti-Affinity is injected properly in leader and workers
  - The new labels are only added if `SubGroupSize` is set
  - The expected TPU environment variables are injected given different combinations of leader requests resources and subgroup sizes. 

> 

Reviewer 2: e2e tests can be added as well if we don't request TPU resources just to test the subgrouping and exclusive placement right?

Author: I added to test the subgrouping, but can e2e be added for exclusive placement? I don't see an existing e2e test for that

Reviewer 2: e2e test means we will really schedule the pods into the nodes. As long as there are more than 1 nodes, then this should be possible right? even if only one nodes, we could still verify the pods have correct labels, and a certain pod is scheduled while others are pending?

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->
To avoid the changes from affecting initial functionality and to not set 
unnecessary labels, there will be no default behavior if `SubGroupSize` is not 
set

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->
### LeaderWorkerSet API
We extend the LeaderWorkerSet API to introduce a new field: subGroupSize to opt in and set the number of pods that each subgroup will contain. Current behavior is kept if not set. 

```
type LeaderWorkerSetSpec struct {
	 // Number of pods per subgroup. This value is immutable,
   // and must not be greater than LeaderWorkerSet.Spec.Size.

> 	SubGroupSize *int32 'json:"subGroupSize,omitempty'

Reviewer 3: Should we put under `leaderWorkerTemplate`, parallel to size? Because it represents the topology of the template.

} 
```

### Exclusive Topology Support

> LeaderWorkerSet can guarantee that the leader and the workers are placed in the same topology if the `leaderworkerset.gke.io/exclusive-topology` annotation is set. Similarly, we will support that the pods within the same subgroup will be placed in the same topology with a new annotation `leaderworkerset.gke.io/exclusive-topology-subgroup`, so that the new changes can support up to two levels of pod affinity. 

Reviewer 3: Avoid to use vendored words, we're now `leaderworkerset.sigs.k8s.io/exclusive-topology`

### Implementation
- Two new annotations will be added

>   - `leaderworkerset.gke.io/exclusive-topology-subgroup `

Reviewer 3: Let's keep the name consistent as word subgroup always **at the head** or **at the end**, head maybe?

Author: Switched it to always be head

  - `leaderworkerset.gke.io/subgroup-size` 
- Three new labels will be added,
  - `leaderworkerset.gke.io/subgroup-index = worker-index/subGroupSize`
    - Tracks which subgroup the pod is part of 
  - `leaderworkerset.gke.io/subgroup-worker-index = worker-index%subGroupSize`
    - index/identity of a pod inside the pod's subgroup
  - `leaderworkerset.gke.io/subgroup-key` 
    - Pods that are part of the same subgroup will have an annotation that is a unique hash value will be generated from the name of the leader, and the subgroup-index

To support exclusive placement at the subgroup level, the pod webhook will inject the new labels, and set the pod affinity/anti-affinity on all the pods. If both levels of pod affinity are set, then the leader pod will contain two pod affinities, while the workers will have a node selector, and a single pod affinity set. 

### TPU Environment Variable Injection

> Because the value of the TPU environmental variables will vary between subgroups, the way they are injected will be extended. `TPU_WORKER_ID` will be the value of `subgroup-worker-index`, while the value of `TPU_HOSTNAMES` will only be a list of the pods in the same subgroup.

Reviewer 3: Do you mean TPU_WORKER_HOSTNAMES rather than TPU_HOSTNAMES

Author: Yes, fixed

LeaderWorkerSet supports the leader not requesting TPU resources. This raises a problem when determining the values of `subgroup-index` and `subgroup-worker-index`. Suppose we have two TPU slices with two hosts each. Since the leader doesn’t request TPU resources, there will be four workers + the leader, meaning that one of the workers will have a worker index of four. Because of the way the new labels are calculated, this worker will have a subgroup-index of two, creating three subgroup indices (0,1,2) even though there are only two TPU slices.

> If the leader does not request TPU resources, then the labels will have the following values

Reviewer 3: Will this lead to vendor problem, because the logic is special as normal. Hope not here.

Author: By vendor problem you mean the logic changes based on whether or not the user uses TPUs?

Reviewer 3: I mean we should not change the label values because of different scenarios, they're identifiers, but  we can leave the logics to the injecting part as we do today. I guess it's possible.

Reviewer 3: This is the last big concern as leaving the TPU logic to accelerators as we do today. see https://github.com/kubernetes-sigs/lws/blob/dde4a46b1db6ce431449ef19c0566000d878ba6e/pkg/utils/accelerators/tpu.go#L105

However, if complete unit test coverage is not possible, explain the reason of it
together with explanation why this is acceptable.
-->

<!--
Additionally, try to enumerate the core package you will be touching
to implement this enhancement and provide the current unit coverage for those
in the form of:
- <package>: <date> - <current test coverage>

This can inform certain test coverage improvements that we want to do before
extending the production code to implement this enhancement.
-->

Will be added to ensure that the annotations are injected into the leader and worker pods, and to verify that the new values of TPU hostnames and TPU worker ID are injected. 

#### Integration tests

<!--
Describe what tests will be added to ensure proper quality of the enhancement.

After the implementation PR is merged, add the names of the tests here.
-->

- The pod-webhook integration tests should test
  - Pod Affinity/Anti-Affinity is injected properly in leader and workers
  - The new labels are only added if `SubGroupSize` is set
  - The expected TPU environment variables are injected given different combinations of leader requests resources and subgroup sizes. 

#### End to End Tests

> - Test that LWS deployment with subgrouping enabled can be deployed

Reviewer 2: ```suggestion
- Test that LWS deployment with subgrouping enabled will have correct pod labels, pod exclusive placement and work well with other features enabled, like failure handling and rolling update.
```
